mh/session.py

```diff
@@ -240,6 +240,15 @@ def get_cities(self):
         return DB.get_cities(self.local_info["server_id"],
                              self.local_info["gate_id"])
 
+    def is_city_empty(self, city_id):
+        return self.get_gate().cities[city_id-1].get_state() == db.LayerState.EMPTY
+
+    def reserve_city(self, city_id, reserve):
+        self.get_gate().cities[city_id-1].reserve(reserve)
+
+    def is_city_reserved(self, city_id):
+        return self.get_gate().cities[city_id-1].reserved
+
```

Comment:
These should have a proper implementation in the database. You might to think `session` as cache that will interact with the database only when necessary. Your current approach is only working because we're using a temporary database that's directly accessible.

```python
    def is_city_empty(self, city_id):
        return DB.get_city(self.local_info["server_id"], self.local_info["gate_id"], city_id).get_state() == db.LayerState.EMPTY

    def reserve_city(self, city_id, reserve):
        return DB.reserve_city(self.local_info["server_id"], self.local_info["gate_id"], city_id, reserve)

    def is_city_reserved(self, city_id):
        return DB.is_city_reserved(self.local_info["server_id"], self.local_info["gate_id"], city_id)
```

I also don't want to expose implementation detail in the `session` module, for instance, `id` starting at 0 internally whereas in the game they usually start at 1.

Reply:
Addressed in latest commit